docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,240 @@
+# User-Defined Type Functions
+
+## Summary
+
+The new type solver for Luau supports a limited number of built-in type functions that enable developers to be expressive with their code. To further enhance this expressivity and the flexibility of the type system, this RFC proposes the design and implementation of user-defined type functions, a feature that will allow developers to define custom type functions.
+
+## Motivation
+
+The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
+- Enable more precise type definitions that can capture complex relationships and constraints
+- Facilitate the creation of reusable and composable libraries
+- Enhance code safety and maintainability with better type-level programming
+
+The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.
+
+## Design
+
+Type functions wil be defined with the following syntax:
+```luau
+type function f(...)
+    -- implementation of the type function
+end
+```
+
+For instance, the `rawget` type function can be written as:
+```luau
+type function rawget(tbl, prop)
```

Comment:
One thing I'm concerned with is that users might write a type function implementations without handling union types. This example is a perfect one: `rawget` is distributive over `tbl` and `prop`, but the implementation doesn't allow for it.

Do we need a `@distributive` attribute for type functions to opt in? Or the inverse `@nondistributive`.

Mostly thinking if there was a way to lift this logic to analysis or compiler that would handle this for the users.

Reply:
Or we could add higher kinded types, `distribute<rawget, { p: "a", q: "x" } | { p: "b", q: "y" }, "p" | "q">`

Reply:
i feel like this should be left as the developers job. we give all the tools they need, but not necessarily help them build whatever they want to build. maybe catching the union case can be left up for the autocomplete?

Reply:
I think it should be comfortable to write the case dealing with unions (e.g. we support iterating over unions), but I don't think we need to be worried about fully automating this stuff away. We ought to wait to see how people use type functions, and what patterns are common and uncommon before we start adding abstractions to optimize writing them.